airbyte-integrations/connectors/destination-iceberg/src/main/java/io/airbyte/integrations/destination/iceberg/IcebergConsumer.java

```diff
@@ -0,0 +1,222 @@
+            if (syncMode == null) {
+                throw new IllegalStateException("Undefined destination sync mode");
+            }
+            final boolean isAppendMode = syncMode != DestinationSyncMode.OVERWRITE;
+            AirbyteStreamNameNamespacePair nameNamespacePair = AirbyteStreamNameNamespacePair.fromAirbyteSteam(stream.getStream());
+            Integer flushBatchSize = catalogConfig.getFormatConfig().getFlushBatchSize();
+            WriteConfig writeConfig = new WriteConfig(namespace, streamName, isAppendMode, flushBatchSize);
+            configs.put(nameNamespacePair, writeConfig);
+            try {
+                spark.sql("DROP TABLE IF EXISTS " + writeConfig.getFullTempTableName());
+            } catch (Exception e) {
+                log.warn("Drop existed temp table failed: {}", e.getMessage(), e);
+            }
+        }
+        this.writeConfigs = configs;
+    }
+
+    /**
+     * call this method when receive a non-STATE AirbyteMessage Ref to <a
+     * href="https://docs.airbyte.com/understanding-airbyte/airbyte-protocol/#airbytemessage">AirbyteMessage</a>
+     */
+    @Override
+    protected void acceptTracked(AirbyteMessage msg) throws Exception {
+        if (msg.getType() != Type.RECORD) {
+            return;
+        }
+        final AirbyteRecordMessage recordMessage = msg.getRecord();
+
+        // ignore other message types.
+        AirbyteStreamNameNamespacePair nameNamespacePair = AirbyteStreamNameNamespacePair.fromRecordMessage(
+            recordMessage);
+        WriteConfig writeConfig = writeConfigs.get(nameNamespacePair);
+        if (writeConfig == null) {
+            throw new IllegalArgumentException(String.format(
+                "Message contained record from a stream that was not in the catalog. namespace: %s , stream: %s",
+                recordMessage.getNamespace(),
+                recordMessage.getStream()));
+        }
+
+        // write data
+        Row row = new GenericRow(new Object[]{UUID.randomUUID().toString(), new Timestamp(recordMessage.getEmittedAt()),
+            Jsons.serialize(recordMessage.getData())});
+        boolean needInsert = writeConfig.addData(row);
+        if (needInsert) {
+            appendToTempTable(writeConfig);
+        }
+    }
+
+    private void appendToTempTable(WriteConfig writeConfig) {
+        String tableName = writeConfig.getFullTempTableName();
+        List<Row> rows = writeConfig.fetchDataCache();
+        //saveAsTable even if rows is empty, to ensure table is created.
+        // otherwise the table would be missing, and throws exception in close()
+        log.info("=> Flushing {} rows into {}", rows.size(), tableName);
+        spark.createDataFrame(rows, normalizationSchema).write()
+            // append data to temp table
+            .mode(SaveMode.Append)
+            //TODO compression config
+            .option("write-format", catalogConfig.getFormatConfig().getFormat().getFormatName()).saveAsTable(tableName);
```

Comment:
Shouldn't Airbyte data records that are received in Json format be transformed to Avro or Parquet format before being written to the warehouse? Or will that be performed by Spark behind the scenes? :)

Reply:
in `acceptTracked()`, airbyte data transform to spark `Row`, and here we declare a file format, then spark will write `Row`s in this format to storage (for Iceberg, wrtie to S3)